### ISSUE TITLE
Stream performance is slightly worse when registering memory under ugni

### ISSUE THREAD

**Issue by ronawho** [chapel-lang/chapel#7852]
When registering the heap under ugni (when a hugepage module is loaded) performance is slightly compared to when no registration is done (or with comm=none or gn-mpi)

Note that dynamic registration (https://github.com/chapel-lang/chapel/pull/6947) significantly improved affinity and stream performance, but we're still slightly behind where we want to be:

| config      | stream-ep perf |
| ----------- | -------------- |
| comm=none   |     ~97.5 GB/s |
| ugni w/ HP  |     ~95.0 GB/s |
| ugni w/o HP |     ~97.5 GB/s |

**Comment by ronawho** [comment]
As I was making this issue, I realized the overhead might be a result of using hugepages . To test this I switched to `CHPL_COMM=none` `CHPL_MEM=cstdlib` `HUGETLB_NO_RESERVE=yes` and loaded a hugepage module (HUGETLB_MORECORE allows libc malloc to grab mem out of hugepage space.) With this config I also see worse performance, so maybe it's just hugepage related and there's nothing we can do about it?


```sh
export CHPL_COMM=none
export CHPL_MEM=cstdlib

chpl test/release/examples/benchmarks/hpcc/stream-ep.chpl --fast --static
./stream 
> 97.5 GB/s

export HUGETLB_NO_RESERVE=yes
module load craype-hugepages2M
./stream 
> 96.0 GB/s
```

That said, I don't quite understand how using hugepages impacts performance here. I suppose that since the pages are bigger, it's likely that ~hugepage_size chunk in the middle of each array that has the wrong affinity, but I'd be surprised a 2M page with the wrong affinity would have that much impact on the perf on ~10GB arrays.

**Comment by ronawho** [comment]
Hmm, I see a similar performance degradation for the reference C+OpenMP version too. ~98 GB/s w/o hugepages, ~96 GB/s w/ 2M hugepages

**Comment by ronawho** [comment]
@gbtitus I'm thinking we can just close this and say, that's the price we pay of using hugepage+first-touch (as opposed to something like multi-ddata, which was able to achieve the full performance). Thoughts?

**Comment by gbtitus** [comment]
It does initially seem surprising that the NUMA chunking due to 2M hugepages would have much effect on performance over 10G arrays.  But performance is measured over an entire `forall-stmt` or equivalent C/C++/Fortan construct then the performance is effectively that of the slowest worker because the timer stops after the sync at the end of the loop.  If the work is split up over 25 cores (for the sake of back-of-the envelope calculations) then each core handles a 400M chunk.  One 2M hugepage is 0.5% of that 400M, so I still wouldn't expect this to result in 2% slower performance, but maybe it's a start at an explanation?  That said, if the array size is exactly divisible by both the hugepage size and the core count then there actually shouldn't be any flaws in the page affinity and we're back to square one on what's going on.  It would be easy to configure the test so that is true, if indeed it isn't already.

**Comment by gbtitus** [comment]
It's possible that the marginal performance loss isn't due to the array references but to secondary effects such as task stacks being not NUMA-local.  I don't know of a way to get a handle on which of a thread's/qthread's/task's references are NUMA-local and which are not.  It's possible the Cray perf tools can tell us this but I don't know them well enough to say.

**Comment by ronawho** [comment]
With `CHPL_COMM=none`, `CHPL_MEM=cstdlib`, `hugepages2M`, `HUGETLB_NO_RESERVE=yes` and `--m=2936012800` (400 2M hugepages per core for a total of ~65.6 GB for all 3 arrays) I still see ~95.5 GB/s

For comm=none, I would expect the task stacks to have good numa affinity still (though I'm not positive of that)

Same config, but with hugepages unloaded, I see ~98 GB/s.

**Comment by gbtitus** [comment]
Remind me: each Qthreads worker gets its own stack pool, right?  If so then yes, with comm=none the task stacks should have the desired NUMA affinity.

**Comment by ronawho** [comment]
Yeah, each worker has its own stack pool

**Comment by gbtitus** [comment]
I don't think the problem could be access to the array metadata -- that will be NUMA non-local for half the cores no matter what the locality granularity is (page v. hugepage).  Plus, it seems to me that only something accessed on every iteration of the innermost kernel can affect the performance, and I believe the metadata references should have been hoisted out of the innermost kernel.

**Comment by ronawho** [comment]
Yeah, metadata should only be accessed once (and wouldn't be an issue for the reference C/OpenMP version)

**Comment by ronawho** [comment]
So trying a hack to just get the arrays allocated with hugepages:

```diff
diff --git a/runtime/include/chpl-mem.h b/runtime/include/chpl-mem.h
index 8f5018d..2b83716 100644
--- a/runtime/include/chpl-mem.h
+++ b/runtime/include/chpl-mem.h
@@ -34,6 +34,7 @@
 #include "chpl-tasks.h"
 #include "error.h"

+#include <hugetlbfs.h>

 /* The names and arguments for these functions are part
    of Chapel's user-facing interface because they are
@@ -127,6 +128,15 @@ void chpl_mem_free(void* memAlloc, int32_t lineno, int32_t filename) {
   chpl_free(memAlloc);
 }

+static inline int alloc_with_hugepage(size_t size) {
+  return (getenv("CHPL_ARRAY_ALLOC_HUGE") != NULL) && size >= gethugepagesize();
+}
+#define ALIGN_DN(i, size)  ((i) & ~((size) - 1))
+#define ALIGN_UP(i, size)  ALIGN_DN((i) + (size) - 1, size)
+static inline size_t hugepage_align(size_t size) {
+  return ALIGN_UP(size, gethugepagesize());
+}
+
 static inline
 void* chpl_mem_array_alloc(size_t nmemb, size_t eltSize, c_sublocid_t subloc,
                            chpl_bool* callAgain, void* repeat_p,
@@ -134,6 +144,11 @@ void* chpl_mem_array_alloc(size_t nmemb, size_t eltSize, c_sublocid_t subloc,
   size_t size = nmemb * eltSize;
   void* p;

+  if (alloc_with_hugepage(size)) {
+    size_t aligned_size = hugepage_align(size);
+    printf("Alloc'ing size of %zd (%zd aligned)\n", size, aligned_size);
+    return get_huge_pages(aligned_size, GHP_DEFAULT);
+  }
   //
   // Temporarily, to support dynamic array registration, we accept a
   // couple of additional arguments.  On a first call, to allocate,
@@ -198,6 +213,11 @@ static inline
 void chpl_mem_array_free(void* p,
                          size_t nmemb, size_t eltSize,
                          int32_t lineno, int32_t filename) {
+  size_t size = nmemb * eltSize;
+  if (alloc_with_hugepage(size)) {
+    free_huge_pages(p);
+    return;
+  }
   if (!chpl_comm_regMemFree(p, nmemb * eltSize))
     chpl_mem_free(p, lineno, filename);
 }
```

With `CHPL_COMM=none`, `CHPL_MEM=jemalloc` (avoid morecore override), `hugepages2M`, `HUGETLB_NO_RESERVE=yes` and `--m=2936012800` I see 98 GB/s (so we know the HUGETLB_ELFMAP and other things don't impact perf)

If I set `CHPL_ARRAY_ALLOC_HUGE=true` I see 95.5 Gb/s

The array alloc will only be used for the raw ddata (not the metadata) so it seems that the perf loss is really just from the arrays being allocated on hugepages

**Comment by gbtitus** [comment]
In the past I've hacked the generated code to check the affinity of the array memory each task works with.  That might be necessary here, as a next step.  One or more tasks must be working with non-NUMA-local array memory.

**Comment by ronawho** [comment]
Yeah, that sounds worthwhile. The affinity must be off somehow, though if it is I'm not sure there's anything we can do about it (since the openmp reference has the same issue, I'm guessing it's something more fundamental)

**Comment by ronawho** [comment]
And fwiw, while I'd like to get to the bottom of this, I don't think it's that high of a priority right now. A ~2% perf loss for stream isn't likely to affect many real codes in a meaningful way. And since it seems to impact the reference version too we can probably ask around and see what internal folks know.
